nvm.sh

```diff
@@ -3687,7 +3687,31 @@ EOF
 }
 
 nvm_supports_xz() {
-  if [ -z "${1-}" ] || ! command which xz >/dev/null 2>&1; then
+  if [ -z "${1-}" ]; then
+    return 1
+  fi
+
+  local NVM_OS
+  NVM_OS="$(nvm_get_os)"
+  if [ "_${NVM_OS}" = '_linux' ]; then
+    if ! command which xz >/dev/null 2>&1; then
+      # Linux without xz on the PATH doesn't support extracting xz-compressed tarballs with tar
+      return 1
+    fi
+  elif [ "_${NVM_OS}" = '_darwin' ]; then
+    local MACOS_VERSION
+    MACOS_VERSION="$(sw_vers -productVersion)"
+    if nvm_version_greater "10.9.0" "${MACOS_VERSION}"; then
+      # macOS 10.8 and earlier doesn't support extracting xz-compressed tarballs with tar
+      return 1
+    fi
+  elif "_${NVM_OS}" = '_freebsd'; then
+    if ! [ -e '/usr/lib/liblzma.so' ]; then
+      # FreeBSD without /usr/lib/liblzma.so doesn't support extracting xz-compressed tarballs with tar
+      return 1
+    fi
+  else
+    # Conservatively assume other operating systems don't support extracting xz-compressed tarballs with tar
     return 1
```

Comment:
is there a reason not to fall back to the current functionality, which is to return 1 only when `command which xz >/dev/null 2>&1`? fails?

Reply:
Practically, no. I'd be comfortable doing that fallback.

Maybe there is some OS out there that behaves different, but I have tested a lot of them, and `xz` on `$PATH` is a pretty reliable indicator for support in `tar`. We handle the big exception to the rule (namely old macOS).

(Taking this to the furthest extent, and being slightly more "YOLO" about it, you could just do the macOS >= 10.9 check on macOS, then fall back to the `xz` on `$PATH` check).

<details><summary>Deeper dive/technical under the hood stuff (click to expand if you want a deeper answer):</summary>

When this wouldn't work: A user could install an OS that doesn't have xz support for `tar` out of the box (such as OpenBSD, or really old OS X/Linux), and install a standalone `xz` binary on the `$PATH` somewhere. Then the code from this PR would over-eagerly ask `tar` to extract xz-compressed tarballs, which would probably fail.

Under the hood, I know that on macOS and BSD there is a shared library actually responsible for xz support in `tar`. _(`/usr/lib/liblzma.dylib` on macOS, `/usr/lib/liblzma.so` on BSD)_. And on older Linux distros, `tar` could be too old to support xz compression.

**_As a coincidence, or by convention, those distros that support xz in `tar` also ship an `xz` binary on the `$PATH`. So it's practically a strong indicator for `tar` xz support._** It's just that if an intepid user installs an xz binary it could fool the simplistic check we do.

I believe Linux and macOS must cover the vast majority of `nvm` users. FreeBSD covers most BSD users, if there are any BSD users out there at all using `nvm`. By reasoning correctly about these, "other users" are so few as to be mostly a theoretical consideration, unless they pop up in the issues of this repo saying something about it.

---

Philosophically: It's "conservative" to fall back to `gzip` just because falling back to `gzip` is expected to work everywhere. Being conservatively wrong and using `gzip` when `xz` could've been used isn't a big deal. On the flip side: being over-optimistically wrong and using `xz` when it doesn't work means the tarball doesn't extract and `nvm` errors out I guess. To be honest, the difference in size between the `gzip` and `xz` tarballs isn't huge... And `gzip` decompresses faster. It's not awful to fall back on `gzip`.

</details>

Leaning toward having all three platform checks then falling back to "xz on `PATH?".

(But we can refactor out the Linux check if you want, since it would be redundant to do the same check twice.)

Reply:
I'm always fine with having a more specific platform check that does a better thing on that platform, but it seems useful to ensure that any machine where it correctly used xz, continues to do so :-)

I think if someone has an `xz` binary that is lying, it's fine if they get an error.

Reply:
Agree. :+1:  Updating this PR when I get a moment.

<details><summary>More technical stuff to support that conclusion</summary>

I just want to clarify that on BSD/macOS, `tar` ignores the `xz` binary and directly links the shared library (`liblzma`) in.

On BSD and macOS it appears the convention is for the `xz` binary to rely on the shared library. If you delete the shared library or rename it, `xz` and `tar -J` break. If you delete or rename the `xz` binary, `tar -J` still works. _**So the `xz` binary is for end-users or for scripts, `tar` doesn't need it at all.**_

But yeah if that's the convention, being on a BSD-style OS and having a standalone `xz` binary that doesn't need `liblzma.[so/dylib]` seems unlikely in any case. Fair to expect it very likely won't happen.

(And of BSDs I tried, only OpenBSD lacks xz support in `tar` so only some fairly niche OSes actually should be detected as "doesn't support xz".)

If they are more like macOS, where the standalone extraction binary for xz isn't `xz` but `gunzip`, then we get a (maybe false) negative and fall back to `gzip`, which is like I said "not awful".

_I have said most or all of this before, but if anyone wants to know without reading the backlog of comments in this PR, this is a pretty compact summary of things._

</details>